File: mini/src/utils/aiEvaluation.ts
import { fbaApi } from '@/api/sdk'
import { useTokenStore } from '@/store'
import { getEnvBaseUrl } from '@/utils'

export interface PracticeAIEvaluationResultPayload {
  reference_answer?: string | null
  grading_summary?: string | null
  strengths?: string[]
  missing_points?: string[]
  improvement_suggestions?: string[]
  encouragement?: string | null
  knowledge_points?: string[]
  needs_manual_review?: boolean
  overview?: string | null
  high_frequency_issues?: string[]
  weak_knowledge_points?: Array<{ name?: string | null, reason?: string | null }>
  next_actions?: string[]
}

export interface PracticeAIEvaluation {
  id: number
  user_id: number
  session_id?: number | null
  practice_record_id?: number | null
  question_id?: number | null
  target_type: 'question_eval' | 'session_summary'
  trigger_source: 'auto' | 'manual'
  status: 'pending' | 'succeeded' | 'failed'
  provider_id?: number | null
  model_name?: string | null
  prompt_version?: string | null
  score?: number | null
  max_score?: number | null
  confidence?: number | null
  summary_text?: string | null
  request_payload?: Record<string, any> | null
  result_payload?: PracticeAIEvaluationResultPayload | null
  error_message?: string | null
  started_at?: string | null
  finished_at?: string | null
  is_latest: boolean
  created_time: string
  updated_time?: string | null
}

interface SubjectiveAnswerOCRResult {
  text: string
}

export function isSubjectiveQuestionType(type?: string | null) {
  return type === 'shortAnswer'
}

export function normalizeStringList(value: unknown) {
  if (!Array.isArray(value))
    return []
  return value.map(item => String(item || '').trim()).filter(Boolean)
}

export function formatEvaluationScore(score?: number | null, maxScore?: number | null) {
  if (score == null && maxScore == null)
    return '--'

  const left = score == null ? '--' : trimTrailingZero(score)
  const right = maxScore == null ? '--' : trimTrailingZero(maxScore)
  return `${left}/${right}`
}

function trimTrailingZero(value: number) {
  const fixed = Number(value)
  if (Number.isInteger(fixed))
    return String(fixed)
  return fixed.toFixed(2).replace(/\.?0+$/, '')
}

export async function getPracticeRecordAIEvaluation(recordId: number) {
  return await fbaApi.qbank.request.get<PracticeAIEvaluation>(`/ai-evaluations/records/${recordId}`)
}

export async function regeneratePracticeRecordAIEvaluation(recordId: number) {
  return await fbaApi.qbank.request.post<PracticeAIEvaluation>(`/ai-evaluations/records/${recordId}/judge`, {
    force_regenerate: true,
  })
}

export async function judgeSessionSubjectiveRecords(sessionId: number, forceRegenerate = false) {
  return await fbaApi.qbank.request.post<PracticeAIEvaluation[]>(
    `/ai-evaluations/sessions/${sessionId}/judge-subjective`,
    { force_regenerate: forceRegenerate },
  )
}

export async function getPracticeSessionAISummary(sessionId: number) {
  return await fbaApi.qbank.request.get<PracticeAIEvaluation>(`/ai-evaluations/sessions/${sessionId}/summary`)
}

export async function generatePracticeSessionAISummary(sessionId: number, forceRegenerate = false) {
  return await fbaApi.qbank.request.post<PracticeAIEvaluation>(
    `/ai-evaluations/sessions/${sessionId}/summary`,
    { force_regenerate: forceRegenerate },
  )
}

export async function recognizeSubjectiveAnswerImage(filePath: string) {
  const tokenStore = useTokenStore()
  const token = await tokenStore.tryGetValidToken()
  if (!token) {
    throw new Error('SUBJECTIVE_OCR_LOGIN_REQUIRED')
  }

  const baseUrl = getEnvBaseUrl().replace(/\/+$/, '')
  const uploadUrl = `${baseUrl}/api/v1/qbank/ai-evaluations/ocr`

  return await new Promise<SubjectiveAnswerOCRResult>((resolve, reject) => {
    uni.uploadFile({
      url: uploadUrl,
      filePath,
      name: 'files',
      header: {
        Authorization: `Bearer ${token}`,
      },
      success: (response) => {
        if (response.statusCode >= 400) {
          try {
            const payload = typeof response.data === 'string'
              ? JSON.parse(response.data || '{}')
              : (response.data as any)
            reject(new Error(payload?.msg || `SUBJECTIVE_OCR_STATUS_${response.statusCode}`))
          }
          catch {
            reject(new Error(`SUBJECTIVE_OCR_STATUS_${response.statusCode}`))
          }
          return
        }

        try {
          const payload = typeof response.data === 'string'
            ? JSON.parse(response.data || '{}')
            : (response.data as any)
          const text = String(payload?.data?.text || '').trim()
          if (!text) {
            reject(new Error(payload?.msg || 'SUBJECTIVE_OCR_EMPTY'))
            return
          }
          resolve({ text })
        }
        catch (error) {
          reject(error)
        }
      },
      fail: reject,
    })
  })
}

File: backend/app/question_bank/schema/ai_evaluation.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
from datetime import datetime
from decimal import Decimal
from typing import Any, Literal

from pydantic import ConfigDict, Field

from backend.common.schema import SchemaBase

PracticeAIEvaluationTargetType = Literal['question_eval', 'session_summary']
PracticeAIEvaluationStatus = Literal['pending', 'succeeded', 'failed']
PracticeAIEvaluationTriggerSource = Literal['auto', 'manual']


class TriggerPracticeAIEvaluationParam(SchemaBase):
    """触发 AI 评估参数"""

    force_regenerate: bool = Field(default=False, description='是否强制重新生成')


class PracticeAIEvaluationRead(SchemaBase):
    """练习 AI 评估结果"""

    model_config = ConfigDict(from_attributes=True)

    id: int = Field(description='评估结果 ID')
    user_id: int = Field(description='用户 ID')
    session_id: int | None = Field(None, description='会话 ID')
    practice_record_id: int | None = Field(None, description='作答记录 ID')
    question_id: int | None = Field(None, description='题目 ID')
    target_type: PracticeAIEvaluationTargetType = Field(description='目标类型')
    trigger_source: PracticeAIEvaluationTriggerSource = Field(description='触发来源')
    status: PracticeAIEvaluationStatus = Field(description='状态')
    provider_id: int | None = Field(None, description='AI 供应商 ID')
    model_name: str | None = Field(None, description='模型名称')
    prompt_version: str | None = Field(None, description='提示词版本')
    score: Decimal | None = Field(None, ge=Decimal('0'), description='得分')
    max_score: Decimal | None = Field(None, ge=Decimal('0'), description='满分')
    confidence: Decimal | None = Field(None, ge=Decimal('0'), le=Decimal('1'), description='置信度')
    summary_text: str | None = Field(None, description='摘要文本')
    request_payload: dict[str, Any] | None = Field(None, description='请求快照')
    result_payload: dict[str, Any] | None = Field(None, description='结果快照')
    error_message: str | None = Field(None, description='错误信息')
    started_at: datetime | None = Field(None, description='开始时间')
    finished_at: datetime | None = Field(None, description='结束时间')
    is_latest: bool = Field(description='是否为最新结果')
    created_time: datetime = Field(description='创建时间')
    updated_time: datetime | None = Field(None, description='更新时间')


class SubjectiveAnswerOCRResult(SchemaBase):
    """主观题 OCR 识别结果"""

    text: str = Field(description='识别后的文本')
